lib/src/lib.rs

```diff
@@ -520,21 +532,25 @@ impl Backend {
 
     pub fn try_connect(&mut self) -> Result<mio::net::TcpStream, ConnectionError> {
         if self.status != BackendStatus::Normal {
-            return Err(ConnectionError::NoBackendAvailable);
+            return Err(ConnectionError::NoBackendAvailable(None));
         }
 
-        //FIXME: what happens if the connect() call fails with EINPROGRESS?
-        let conn = mio::net::TcpStream::connect(self.address)
-            .map_err(|_| ConnectionError::NoBackendAvailable);
-        if conn.is_ok() {
-            //self.retry_policy.succeed();
-            self.inc_connections();
-        } else {
-            self.retry_policy.fail();
-            self.failures += 1;
+        match mio::net::TcpStream::connect(self.address) {
+            Ok(tcp_stream) => {
+                //self.retry_policy.succeed();
+                self.inc_connections();
+                Ok(tcp_stream)
+            }
+            Err(_connection_error) => {
+                self.retry_policy.fail();
+                self.failures += 1;
+                // TODO: handle EINPROGRESS. It is difficult. It is discussed here:
+                // https://docs.rs/mio/latest/mio/net/struct.TcpStream.html#method.connect
+                // with an example code here:
+                // https://github.com/Thomasdezeeuw/heph/blob/0c4f1ab3eaf08bea1d65776528bfd6114c9f8374/src/net/tcp/stream.rs#L560-L622
+                Err(ConnectionError::NoBackendAvailable(None))
```

Comment:
Is it really that there is no backend available or that we are unable to create a connection ?

Reply:
Indeed you're right ! I created the `MioConnectError` variant for this occasion.